File: Tetris3D.java
// Tetris3D.java - 3D Тетрис на Java
import org.jline.terminal.Terminal;
import org.jline.terminal.TerminalBuilder;
import org.jline.utils.InfoCmp;
import com.google.gson.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Tetris3D {
    private static final int WIDTH = 6;
    private static final int DEPTH = 6;
    private static final int HEIGHT = 6;
    private static final int EMPTY = 0;
    private static final String RECORD_FILE = "tetris3d_record.json";

    private static final Map<String, int[][]> SHAPES = new HashMap<>();
    static {
        SHAPES.put("I", new int[][]{{0,0,0},{1,0,0},{2,0,0},{3,0,0}});
        SHAPES.put("O", new int[][]{{0,0,0},{1,0,0},{0,1,0},{1,1,0}});
        SHAPES.put("T", new int[][]{{0,0,0},{1,0,0},{2,0,0},{1,1,0}});
        SHAPES.put("L", new int[][]{{0,0,0},{1,0,0},{2,0,0},{2,1,0}});
        SHAPES.put("J", new int[][]{{0,0,0},{1,0,0},{2,0,0},{0,1,0}});
        SHAPES.put("S", new int[][]{{1,0,0},{2,0,0},{0,1,0},{1,1,0}});
        SHAPES.put("Z", new int[][]{{0,0,0},{1,0,0},{1,1,0},{2,1,0}});
    }

    private Terminal terminal;
    private int[][][] field;
    private int score, speed;
    private boolean gameOver, paused, running;
    private int record;
    private List<int[]> currentShape;
    private int[] shapePos;
    private long dropTimer;
    private boolean lock;

    public Tetris3D() throws IOException {
        terminal = TerminalBuilder.builder().system(true).build();
        resetState();
        record = loadRecord();
        running = true;
        lock = false;
    }

    private void resetState() {
        field = new int[HEIGHT][WIDTH][DEPTH];
        score = 0;
        speed = 1;
        gameOver = false;
        paused = false;
        currentShape = null;
        shapePos = new int[]{0,0,0};
        dropTimer = System.currentTimeMillis();
    }

    private int loadRecord() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(RECORD_FILE)));
            JsonObject obj = new Gson().fromJson(content, JsonObject.class);
            return obj.get("record").getAsInt();
        } catch (Exception e) { return 0; }
    }

    private void saveRecord() {
        try {
            JsonObject obj = new JsonObject();
            obj.addProperty("record", record);
            Files.write(Paths.get(RECORD_FILE), obj.toString().getBytes());
        } catch (Exception e) {}
    }

    private List<int[]> randomShape() {
        String[] names = SHAPES.keySet().toArray(new String[0]);
        String name = names[new Random().nextInt(names.length)];
        int[][] coords = SHAPES.get(name);
        List<int[]> shape = new ArrayList<>();
        for (int[] p : coords) shape.add(p.clone());
        return shape;
    }

    private boolean spawnShape() {
        List<int[]> shape = randomShape();
        int[] xs = shape.stream().mapToInt(p -> p[0]).toArray();
        int[] zs = shape.stream().mapToInt(p -> p[1]).toArray();
        int cx = (Arrays.stream(xs).max().getAsInt() + Arrays.stream(xs).min().getAsInt()) / 2;
        int cz = (Arrays.stream(zs).max().getAsInt() + Arrays.stream(zs).min().getAsInt()) / 2;
        int startX = WIDTH/2 - cx;
        int startZ = DEPTH/2 - cz;
        int startY = HEIGHT - 1;
        for (int[] p : shape) {
            int x = startX + p[0], z = startZ + p[1], y = startY + p[2];
            if (x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT || field[y][x][z] != EMPTY) {
                gameOver = true;
                if (score > record) { record = score; saveRecord(); }
                return false;
            }
        }
        currentShape = shape;
        shapePos = new int[]{startX, startZ, startY};
        return true;
    }

    private boolean collides(List<int[]> shape, int[] pos) {
        for (int[] p : shape) {
            int x = pos[0] + p[0], z = pos[1] + p[1], y = pos[2] + p[2];
            if (x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT) return true;
            if (field[y][x][z] != EMPTY) return true;
        }
        return false;
    }

    private void lockShape() {
        for (int[] p : currentShape) {
            int x = shapePos[0] + p[0], z = shapePos[1] + p[1], y = shapePos[2] + p[2];
            field[y][x][z] = 1;
        }
        int cleared = 0;
        for (int y = 0; y < HEIGHT; y++) {
            boolean full = true;
            for (int x = 0; x < WIDTH; x++) for (int z = 0; z < DEPTH; z++) if (field[y][x][z] == EMPTY) { full = false; break; }
            if (full) {
                for (int y2 = y; y2 > 0; y2--) field[y2] = field[y2-1];
                field[0] = new int[WIDTH][DEPTH];
                cleared++;
            }
        }
        if (cleared > 0) { score += cleared; speed = 1 + score/3; }
        if (!spawnShape()) gameOver = true;
    }

    private boolean move(int dx, int dz, int dy) {
        int[] newPos = {shapePos[0]+dx, shapePos[1]+dz, shapePos[2]+dy};
        if (!collides(currentShape, newPos)) { shapePos = newPos; return true; }
        return false;
    }

    private void rotate() {
        List<int[]> newShape = new ArrayList<>();
        for (int[] p : currentShape) newShape.add(new int[]{-p[1], p[0], p[2]});
        if (!collides(newShape, shapePos)) currentShape = newShape;
    }

    private void drop() {
        while (move(0,0,-1)) {}
        lockShape();
    }

    private void update() {
        if (gameOver || paused) return;
        if (System.currentTimeMillis() - dropTimer >= 1000 / speed) {
            if (!move(0,0,-1)) lockShape();
            dropTimer = System.currentTimeMillis();
        }
    }

    private int[][][] getFieldWithShape() {
        int[][][] copy = new int[HEIGHT][WIDTH][DEPTH];
        for (int y = 0; y < HEIGHT; y++) for (int x = 0; x < WIDTH; x++) copy[y][x] = field[y][x].clone();
        if (currentShape != null) {
            for (int[] p : currentShape) {
                int x = shapePos[0]+p[0], z = shapePos[1]+p[1], y = shapePos[2]+p[2];
                if (x>=0 && x<WIDTH && z>=0 && z<DEPTH && y>=0 && y<HEIGHT) copy[y][x][z] = 2;
            }
        }
        return copy;
    }

    private void draw() {
        terminal.puts(InfoCmp.Capability.clear_screen);
        int[][][] fieldCopy = getFieldWithShape();
        System.out.println("═".repeat(WIDTH*2+10));
        System.out.printf("  Счёт: %d   Рекорд: %d   Скорость: %d%n", score, record, speed);
        System.out.println("═".repeat(WIDTH*2+10));
        for (int y = HEIGHT-1; y >= 0; y--) {
            System.out.printf("  Y=%d  ", y);
            for (int z = 0; z < DEPTH; z++) {
                for (int x = 0; x < WIDTH; x++) {
                    int val = fieldCopy[y][x][z];
                    if (val == 2) System.out.print("\u001B[32m█ \u001B[0m");
                    else if (val == 1) System.out.print("\u001B[36m█ \u001B[0m");
                    else System.out.print("· ");
                }
                System.out.print("  ");
            }
            System.out.println();
        }
        System.out.println("═".repeat(WIDTH*2+10));
        String status = paused ? "ПАУЗА" : gameOver ? "ИГРА ОКОНЧЕНА" : "ИГРА";
        System.out.printf("  %s  |  ←/→ - X  |  W/S - Z  |  Пробел - вращение  |  ↓ - падение%n", status);
        System.out.println("  P - пауза  |  R - рестарт  |  Q - выход");
    }

    private void reset() {
        resetState();
        spawnShape();
    }

    private void handleInput() {
        try {
            while (running) {
                int ch = terminal.reader().read();
                if (ch == -1) continue;
                if (lock) continue;
                lock = true;
                char c = (char) ch;
                switch (c) {
                    case 'a': case 'A': if (!gameOver && !paused) move(-1,0,0); break;
                    case 'd': case 'D': if (!gameOver && !paused) move(1,0,0); break;
                    case 'w': case 'W': if (!gameOver && !paused) move(0,-1,0); break;
                    case 's': case 'S': if (!gameOver && !paused) move(0,1,0); break;
                    case ' ': rotate(); break;
                    case '↓': drop(); break;
                    case 'p': case 'P': paused = !paused; break;
                    case 'r': case 'R': reset(); break;
                    case 'q': case 'Q': running = false; break;
                }
                lock = false;
            }
        } catch (IOException e) {}
    }

    public void run() throws Exception {
        spawnShape();
        Thread inputThread = new Thread(this::handleInput);
        inputThread.setDaemon(true);
        inputThread.start();

        while (running) {
            update();
            draw();
            Thread.sleep(1000/30);
        }
        terminal.close();
    }

    public static void main(String[] args) throws Exception {
        Tetris3D game = new Tetris3D();
        game.run();
    }
}

File: tetris3d.go
// tetris3d.go - 3D Тетрис на Go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"time"

	"github.com/eiannone/keyboard"
)

const (
	WIDTH  = 6
	DEPTH  = 6
	HEIGHT = 6
	EMPTY  = 0
	RECORD_FILE = "tetris3d_record.json"
)

var SHAPES = map[string][][3]int{
	"I": {{0,0,0},{1,0,0},{2,0,0},{3,0,0}},
	"O": {{0,0,0},{1,0,0},{0,1,0},{1,1,0}},
	"T": {{0,0,0},{1,0,0},{2,0,0},{1,1,0}},
	"L": {{0,0,0},{1,0,0},{2,0,0},{2,1,0}},
	"J": {{0,0,0},{1,0,0},{2,0,0},{0,1,0}},
	"S": {{1,0,0},{2,0,0},{0,1,0},{1,1,0}},
	"Z": {{0,0,0},{1,0,0},{1,1,0},{2,1,0}},
}

type Tetris3D struct {
	field        [][][]int
	score        int
	speed        int
	gameOver     bool
	paused       bool
	running      bool
	record       int
	currentShape [][3]int
	shapePos     [3]int // x,z,y
	dropTimer    time.Time
	lockInput    bool
	rand         *rand.Rand
}

func NewTetris3D() *Tetris3D {
	t := &Tetris3D{
		field:    make([][][]int, HEIGHT),
		score:    0,
		speed:    1,
		gameOver: false,
		paused:   false,
		running:  true,
		record:   loadRecord(),
		rand:     rand.New(rand.NewSource(time.Now().UnixNano())),
	}
	for y := 0; y < HEIGHT; y++ {
		t.field[y] = make([][]int, WIDTH)
		for x := 0; x < WIDTH; x++ {
			t.field[y][x] = make([]int, DEPTH)
		}
	}
	t.spawnShape()
	return t
}

func loadRecord() int {
	file, err := os.Open(RECORD_FILE)
	if err != nil {
		return 0
	}
	defer file.Close()
	var data map[string]int
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&data); err != nil {
		return 0
	}
	if val, ok := data["record"]; ok {
		return val
	}
	return 0
}

func saveRecord(record int) {
	data := map[string]int{"record": record}
	file, _ := os.Create(RECORD_FILE)
	defer file.Close()
	encoder := json.NewEncoder(file)
	encoder.Encode(data)
}

func (t *Tetris3D) randomShape() [][3]int {
	keys := make([]string, 0, len(SHAPES))
	for k := range SHAPES {
		keys = append(keys, k)
	}
	name := keys[t.rand.Intn(len(keys))]
	return SHAPES[name]
}

func (t *Tetris3D) spawnShape() bool {
	shape := t.randomShape()
	xs := make([]int, len(shape))
	zs := make([]int, len(shape))
	for i, p := range shape {
		xs[i] = p[0]
		zs[i] = p[1]
	}
	cx := (max(xs) + min(xs)) / 2
	cz := (max(zs) + min(zs)) / 2
	startX := WIDTH/2 - cx
	startZ := DEPTH/2 - cz
	startY := HEIGHT - 1
	for _, p := range shape {
		x := startX + p[0]
		z := startZ + p[1]
		y := startY + p[2]
		if x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT || t.field[y][x][z] != EMPTY {
			t.gameOver = true
			if t.score > t.record {
				t.record = t.score
				saveRecord(t.record)
			}
			return false
		}
	}
	t.currentShape = shape
	t.shapePos = [3]int{startX, startZ, startY}
	return true
}

func (t *Tetris3D) collides(shape [][3]int, pos [3]int) bool {
	for _, p := range shape {
		x := pos[0] + p[0]
		z := pos[1] + p[1]
		y := pos[2] + p[2]
		if x < 0 || x >= WIDTH || z < 0 || z >= DEPTH || y < 0 || y >= HEIGHT {
			return true
		}
		if t.field[y][x][z] != EMPTY {
			return true
		}
	}
	return false
}

func (t *Tetris3D) lockShape() {
	for _, p := range t.currentShape {
		x := t.shapePos[0] + p[0]
		z := t.shapePos[1] + p[1]
		y := t.shapePos[2] + p[2]
		t.field[y][x][z] = 1
	}
	cleared := 0
	for y := 0; y < HEIGHT; y++ {
		full := true
		for x := 0; x < WIDTH; x++ {
			for z := 0; z < DEPTH; z++ {
				if t.field[y][x][z] == EMPTY {
					full = false
					break
				}
			}
			if !full {
				break
			}
		}
		if full {
			for y2 := y; y2 > 0; y2-- {
				t.field[y2] = t.field[y2-1]
			}
			t.field[0] = make([][]int, WIDTH)
			for x := 0; x < WIDTH; x++ {
				t.field[0][x] = make([]int, DEPTH)
			}
			cleared++
		}
	}
	if cleared > 0 {
		t.score += cleared
		t.speed = 1 + t.score/3
	}
	if !t.spawnShape() {
		t.gameOver = true
	}
}

func (t *Tetris3D) move(dx, dz, dy int) bool {
	newPos := [3]int{t.shapePos[0] + dx, t.shapePos[1] + dz, t.shapePos[2] + dy}
	if !t.collides(t.currentShape, newPos) {
		t.shapePos = newPos
		return true
	}
	return false
}

func (t *Tetris3D) rotate() {
	newShape := make([][3]int, len(t.currentShape))
	for i, p := range t.currentShape {
		newShape[i] = [3]int{-p[1], p[0], p[2]} // вращение вокруг Y
	}
	if !t.collides(newShape, t.shapePos) {
		t.currentShape = newShape
	}
}

func (t *Tetris3D) drop() {
	for t.move(0, 0, -1) {
	}
	t.lockShape()
}

func (t *Tetris3D) update() {
	if t.gameOver || t.paused {
		return
	}
	if time.Since(t.dropTimer) >= time.Second/time.Duration(t.speed) {
		if !t.move(0, 0, -1) {
			t.lockShape()
		}
		t.dropTimer = time.Now()
	}
}

func (t *Tetris3D) getFieldWithShape() [][][]int {
	copyField := make([][][]int, HEIGHT)
	for y := 0; y < HEIGHT; y++ {
		copyField[y] = make([][]int, WIDTH)
		for x := 0; x < WIDTH; x++ {
			copyField[y][x] = make([]int, DEPTH)
			copy(copyField[y][x], t.field[y][x])
		}
	}
	if t.currentShape != nil {
		for _, p := range t.currentShape {
			x := t.shapePos[0] + p[0]
			z := t.shapePos[1] + p[1]
			y := t.shapePos[2] + p[2]
			if x >= 0 && x < WIDTH && z >= 0 && z < DEPTH && y >= 0 && y < HEIGHT {
				copyField[y][x][z] = 2
			}
		}
	}
	return copyField
}

func (t *Tetris3D) draw() {
	clearScreen()
	fieldCopy := t.getFieldWithShape()
	fmt.Println(stringRepeat("═", WIDTH*2+10))
	fmt.Printf("  Счёт: %d   Рекорд: %d   Скорость: %d\n", t.score, t.record, t.speed)
	fmt.Println(stringRepeat("═", WIDTH*2+10))
	for y := HEIGHT - 1; y >= 0; y-- {
		fmt.Printf("  Y=%d  ", y)
		for z := 0; z < DEPTH; z++ {
			for x := 0; x < WIDTH; x++ {
				val := fieldCopy[y][x][z]
				if val == 2 {
					fmt.Print("\x1b[32m█ \x1b[0m")
				} else if val == 1 {
					fmt.Print("\x1b[36m█ \x1b[0m")
				} else {
					fmt.Print("· ")
				}
			}
			fmt.Print("  ")
		}
		fmt.Println()
	}
	fmt.Println(stringRepeat("═", WIDTH*2+10))
	status := "ИГРА"
	if t.paused {
		status = "ПАУЗА"
	} else if t.gameOver {
		status = "ИГРА ОКОНЧЕНА"
	}
	fmt.Printf("  %s  |  ←/→ - X  |  W/S - Z  |  Пробел - вращение  |  ↓ - падение\n", status)
	fmt.Println("  P - пауза  |  R - рестарт  |  Q - выход")
}

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func stringRepeat(s string, n int) string {
	res := ""
	for i := 0; i < n; i++ {
		res += s
	}
	return res
}

func min(arr []int) int {
	m := arr[0]
	for _, v := range arr[1:] {
		if v < m {
			m = v
		}
	}
	return m
}

func max(arr []int) int {
	m := arr[0]
	for _, v := range arr[1:] {
		if v > m {
			m = v
		}
	}
	return m
}

func (t *Tetris3D) reset() {
	t.field = make([][][]int, HEIGHT)
	for y := 0; y < HEIGHT; y++ {
		t.field[y] = make([][]int, WIDTH)
		for x := 0; x < WIDTH; x++ {
			t.field[y][x] = make([]int, DEPTH)
		}
	}
	t.score = 0
	t.speed = 1
	t.gameOver = false
	t.paused = false
	t.spawnShape()
}

func (t *Tetris3D) handleInput() {
	for t.running {
		char, key, err := keyboard.GetKey()
		if err != nil {
			continue
		}
		if t.lockInput {
			continue
		}
		t.lockInput = true
		switch key {
		case keyboard.KeyArrowLeft:
			if !t.gameOver && !t.paused {
				t.move(-1, 0, 0)
			}
		case keyboard.KeyArrowRight:
			if !t.gameOver && !t.paused {
				t.move(1, 0, 0)
			}
		case keyboard.KeyArrowDown:
			if !t.gameOver && !t.paused {
				t.drop()
			}
		case keyboard.KeySpace:
			if !t.gameOver && !t.paused {
				t.rotate()
			}
		default:
			switch char {
			case 'w', 'W':
				if !t.gameOver && !t.paused {
					t.move(0, -1, 0)
				}
			case 's', 'S':
				if !t.gameOver && !t.paused {
					t.move(0, 1, 0)
				}
			case 'p', 'P':
				t.paused = !t.paused
			case 'r', 'R':
				t.reset()
			case 'q', 'Q':
				t.running = false
				return
			}
		}
		t.lockInput = false
	}
}

func (t *Tetris3D) run() {
	go t.handleInput()

	t.dropTimer = time.Now()
	ticker := time.NewTicker(time.Second / 30)
	defer ticker.Stop()

	for t.running {
		<-ticker.C
		t.update()
		t.draw()
	}
}

func main() {
	game := NewTetris3D()
	game.run()
	fmt.Println("Игра завершена.")
}
